### PR TITLE
Add a PhoneNumber class to handle formatting and actions on numbers

### PR DESCRIPTION
**Pull request by caitlinrhd**
We value collaboration at Invoca, and focus on peer mentorship as a way for the whole team to learn and thrive. As part of this process, we conduct code reviews during our development cycle. In this phase of the interview, we would like you to review and comment on this Pull Request. You can assume that this Pull Request was written by an intern, who is new to professional software development.  Please take about 20-30 minutes to review this Pull Request and comment as if you were reviewing an intern's code.

### CODE REVIEW

File: phone_actions.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+class PhoneActions
+
+  def initialize(username, password)
+    @actions_url = "https://#{username}:#{password}@api.<made_up>.com/actions/"
+  end
+
+  def call!(number, message)
```
Comment (lech-kopec) [inline phone_actions.rb:7]: As part of the ruby/rails convention, we try to give ! mark to method's name only when it's considered dangerous. For example when method is modifying object its called on. Checkout difference between upcase and upcase! on strings

File: phone_actions.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    JSON.parse(RestClient.post("@actions_url/#{text}", {message: message, number: number}))
```
Comment (lech-kopec) [inline phone_actions.rb:8]: Couple points here:

- Looks like you're missing string evaluation braces for action_url. I think it should be wrapped with #{}
- Also, looks like your actions_url already ends with / so you don't have add one more
- I might be missing something here, but I don't see where is variable text coming from. Are you missing one input paramter?
- I'm not familiar with RestClient so please correct me if I'm wrong, but I think you shouldn't everything that's return from post method. We're probably only interested in body.
- Before you get to parsing body, it would be great to check the status of the response and handle failed requests. Have you checked in documentation what is the expected behavior when call fails? Are we supposed to just display error or we should retry couple times?

File: phone_actions.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  end
+
+  def text!(number, message)
+    JSON.parse(RestClient.post("@actions_url/#{call}", {message: message, number: number}))
```
Comment (lech-kopec) [inline phone_actions.rb:12]: All the comments from previous method also apply here.
Additionally, you probably noticed that we have some code repetition here - json parsing and constructing the call. While json parsing repetition is not a big deal, it would be much nicer to rest call construction in some helper method. This way, your code won't have to worry about how body should look like and if we decide in future to update the body, we will have to update only one place.

File: phone_actions.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
+class PhoneActions
+
+  def initialize(username, password)
+    @actions_url = "https://#{username}:#{password}@api.<made_up>.com/actions/"
```
Comment (lech-kopec) [inline phone_actions.rb:4]: Is the url address always the same? I'd assume it will be different for each environment so we might have to move this out to config file and import it from there.

File: phone_number.rb [hunk 5]
```diff
@@ -0,0 +1,63 @@
+require_relative "./phone_actions"
```
Comment (lech-kopec) [inline phone_number.rb:1]: I think you can just do
require_relative "phone_actions" so without ./

File: phone_number.rb [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+class PhoneNumber
+  USERNAME = ENV['ACTIONS_USERNAME']
+  PASSWORD = ENV['ACTIONS_PASSWORD']
```
Comment (lech-kopec) [inline phone_number.rb:5]: Let's extract fetching credentials to separate class/module.
In our code, we don't want to worry about where the credentials are coming from and what are the prop names.
Having separate, dedicated module will also help us in case when we would have to switch how we get those - change will have to happen only in one place.

File: phone_number.rb [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+  def initialize(number)
+    @number = number
+    @number = format_number
+  end
```
Comment (lech-kopec) [inline phone_number.rb:10]: Let's make format_number accept our input argument and then initialize our instance variable simply by
@number = format_number(number)

File: phone_number.rb [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+  def call(message)
+    PhoneActions.new(USERNAME, PASSWORD).call!(@number, message)
+  end
+
+  def text_and_call(message)
+    PhoneActions.new(USERNAME, PASSWORD).text!(@number, message)
+    PhoneActions.new(USERNAME, PASSWORD).call!(message, message)
+  end
```
Comment (lech-kopec) [inline phone_number.rb:23]: Right now, there is a lot of duplication here, so before I give any comments let's see how this will looks like after you move out the credentials.
One thing I can mention here already is that since you already have text and call methods defined, you can just use them inside text_and_call (that will make our code more DRY and will help reduce bugs, one of which we can already see on like 22 where you provide message as both arguments to the call method)

File: phone_number.rb [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+  def number
```
Comment (lech-kopec) [inline phone_number.rb:25]: Let's remove this and define attribute reader for it (using attr_reader)

File: phone_number.rb [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    @number
+  end
+
+  def to_i
+    n = @number.gsub!(/[^0-9]/, "")
```
Comment (lech-kopec) [inline phone_number.rb:30]: Always be careful with methods that end with ! It's always better to triple check them.
gsub with ! will modify object it's executed on, so your instance variable will be modified (in-place) and I'm assuming that's not what you want.
Additionally, you can write this in just one line, without intermediate variable

File: phone_number.rb [hunk 11]
```diff
@@ -0,0 +1,63 @@
+require_relative "./phone_actions"
+
+class PhoneNumber
+  USERNAME = ENV['ACTIONS_USERNAME']
+  PASSWORD = ENV['ACTIONS_PASSWORD']
+
+  def initialize(number)
+    @number = number
```
Comment (lech-kopec) [inline phone_number.rb:8]: It's always good idea to check the input. Here we would like to make sure that number is not nil and that it's actually a string.

File: phone_number.rb [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    n.to_i
+  end
+
+  private
+
+  def format_number
+    "(#{find_area_code}) #{find_exchange}-#{find_line_number}"
+  end
+
+  def find_area_code
```
Comment (lech-kopec) [inline phone_number.rb:40]: There are couple of issues here (use of ! method on string, using magic-numbers to access substring) so maybe it would be cleaner to solve this with regex matchers

File: phone_number_spec.rb [hunk 13]
```diff
@@ -0,0 +1,41 @@
+require_relative "./phone_number"
+
+RSpec.describe PhoneNumber do
```
Comment (lech-kopec) [inline phone_number_spec.rb:3]: let's add here more scenarios where check corner cases, right now, it's just one happy-path.
It would be great to also write unite tests for phone_actions

File: phone_number_spec.rb [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+    it "is the number as an int" do
+      new_number = PhoneNumber.new("555-867-5309")
+      expect(new_number.to_i).to eq(5558675309)
+    end
+  end
+
+  describe "#text" do
+    it "calls PhoneActions#text!" do
+      expect_any_instance_of(PhoneActions).to receive(:text!)
+      phone_number.text("hello world!")
```
Comment (lech-kopec) [inline phone_number_spec.rb:22]: Please check also for expected outputs from the method

File: phone_number_spec.rb [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+  describe "#to_i" do
+    it "is the number as an int" do
+      new_number = PhoneNumber.new("555-867-5309")
+      expect(new_number.to_i).to eq(5558675309)
+    end
+  end
+
+  describe "#text" do
+    it "calls PhoneActions#text!" do
+      expect_any_instance_of(PhoneActions).to receive(:text!)
```
Comment (lech-kopec) [inline phone_number_spec.rb:21]: We should be also checking input parameters to the method and returned value

**Review by lech-kopec — CHANGES_REQUESTED** [review]
It's a good start, let's work together to implement improvements to make more secure and DRY.
Reach out to me if you have any questions.
